README.md

A package to exhibit your work portfolio/products catalogue with rolling images.

## Getting started
## Usage

```dart
import 'package:rolling_images/rolling_images.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Rolling Images',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: RollingCards(
        viewportFraction: 1,
        imageUrls: const [
          'https://images.pexels.com/photos/11730329/pexels-photo-11730329.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1',
          'https://images.pexels.com/photos/2295744/pexels-photo-2295744.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1',
          'https://images.pexels.com/photos/3777622/pexels-photo-3777622.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1',
          'https://images.pexels.com/photos/3186010/pexels-photo-3186010.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1',
        ],
        isLocalAsset: false,
        onPageChanged: (int page) {
          // Get Notified about current page
        },
      ),
    );
  }
}

```
## Features
## App Screenshots

<table>
  <tr>
    <td>BoxShape = Rectangle</td>
     <td>BoxShape = Circle</td>
  </tr>
  <tr>
    <td><img src="https://raw.githubusercontent.com/devAnkitGupta/rolling_images/master/assets/images/rollingImageRect.gif"></td>
    <td><img src="https://raw.githubusercontent.com/devAnkitGupta/rolling_images/master/assets/images/rollingImageCircle.gif"></td>
  </tr>
 </table>

## Web Screenshots

<img src='https://raw.githubusercontent.com/devAnkitGupta/rolling_images/master/assets/images/rollingImageWeb.gif'>



CHANGELOG.md
## 0.0.1
Initial Release

## 0.0.2
Adding Documentation 

lib/src/rolling_cards.dart
import 'package:flutter/material.dart';
import './rolling_container.dart';

/// Rolling Cards
class RollingCards extends StatefulWidget {
  const RollingCards({
    Key? key,
    required this.imageUrls,
    required this.isLocalAsset,
    this.viewportFraction = 0.96,
    this.borderRadius,
    this.imageContainerShape,
    this.fit,
    this.margin,
    this.onPageChanged,
  }) : super(key: key);

  /// List of urls (local/network) which need to be displayed
  final List<String> imageUrls;

  /// flage by which widget identifies wheather the image are local
  /// assets or needed to be fetched from users
  final bool isLocalAsset;

  /// The fraction of the viewport that each page should occupy.
  /// Defaults to 1.0, which means each image page fills the viewport in the scrolling direction.
  final double viewportFraction;

  /// BorderRadius for image default value is 20
  /// provide 0 for no border radius
  final double? borderRadius;

  /// The shape to use when rendering a image.
  /// Default value is Rectangle
  final BoxShape? imageContainerShape;

  /// How a box should be inscribed into another box.
  final BoxFit? fit;

  /// Margins provided for image
  ///  Default is
  ///
  /// ``` dart
  /// const EdgeInsets.only(
  /// top: 20,
  /// bottom: 20,
  /// left: 15,
  /// right: 15,
  /// )
  /// ```
  final EdgeInsets? margin;

  /// Callback returning current image display value
  /// initial image value when build is 0
  final Function(int)? onPageChanged;

  @override
  _RollingCardsState createState() => _RollingCardsState();
}

class _RollingCardsState extends State<RollingCards> {
  late PageController _controller;
  double currentPage = 0.0;
  Duration animateTime = const Duration(milliseconds: 700);
  Curve curve = Curves.ease;
  bool isAnimating = false;

  /// Convert List of urls to Map<int, String>
  /// to produce keys with index of itself in list
  late Map<int, String> indexedMapUrl;

  @override
  void initState() {
    _controller = PageController(
      viewportFraction: widget.viewportFraction,
      keepPage: true,
    );
    indexedMapUrl = widget.imageUrls.asMap();
    super.initState();
  }

  @override
  void didChangeDependencies() {
    _controller.addListener(
      () {
        Future.microtask(
          () => setState(
            () {
              currentPage = _controller.page!;
            },
          ),
        );
      },
    );
    super.didChangeDependencies();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return PageView(
      controller: _controller,
      children: indexedMapUrl.entries
          .map(
            (e) => ImageCard(
              idx: e.key,
              currentPage: currentPage,
              key: Key('${e.key} ${e.value}'),
              url: e.value,
              isLocalImage: widget.isLocalAsset,
              borderRadius: widget.borderRadius ?? 20,
              boxShape: widget.imageContainerShape ?? BoxShape.rectangle,
              fit: widget.fit ?? BoxFit.cover,
              margin: widget.margin ??
                  const EdgeInsets.only(
                    top: 20,
                    bottom: 20,
                    left: 15,
                    right: 15,
                  ),
            ),
          )
          .toList(),
      onPageChanged: (value) {
        if (widget.onPageChanged != null) {
          widget.onPageChanged!(value);
        }
      },
    );
  }
}


lib/src/rolling_container.dart
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';

class ImageCard extends StatelessWidget {
  /// Index of Image
  final int idx;

  /// Current page in mainView
  final double currentPage;

  /// Address of the image
  final String url;

  /// How a box should be inscribed into another box.
  final BoxFit fit;
  final EdgeInsets margin;
  final double borderRadius;
  final bool isLocalImage;
  final BoxShape boxShape;

  const ImageCard({
    Key? key,
    required this.idx,
    required this.currentPage,
    required this.url,
    required this.fit,
    required this.margin,
    required this.borderRadius,
    required this.isLocalImage,
    required this.boxShape,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    double relativePosition = idx - currentPage;

    /// Transforming the child images.
    return Transform(
      transform: Matrix4.identity()
        ..setEntry(3, 2, 0.002)
        ..scale((1 - relativePosition.abs()).clamp(0.2, 0.6) + 0.4)
        ..rotateZ(relativePosition),
      alignment:
          relativePosition >= 0 ? Alignment.centerLeft : Alignment.centerRight,
      child: Container(
        margin: margin,
        decoration: BoxDecoration(
          borderRadius: boxShape == BoxShape.circle
              ? null
              : BorderRadius.circular(borderRadius),
          image: isLocalImage
              ? DecorationImage(image: AssetImage(url), fit: fit)
              : DecorationImage(
                  image: CachedNetworkImageProvider(url, cacheKey: url),
                  fit: fit,
                ),
          shape: boxShape,
        ),
      ),
    );
  }
}


example/lib/example.dart
import 'package:flutter/material.dart';
import 'package:rolling_images/rolling_images.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Rolling Images',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: RollingCards(
        viewportFraction: 1,
        imageUrls: const [
          'https://images.pexels.com/photos/11730329/pexels-photo-11730329.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1',
          'https://images.pexels.com/photos/2295744/pexels-photo-2295744.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1',
          'https://images.pexels.com/photos/3777622/pexels-photo-3777622.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1',
          'https://images.pexels.com/photos/3186010/pexels-photo-3186010.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1',
        ],
        isLocalAsset: false,
        onPageChanged: (int page) {
          // Get Notified about current page
        },
      ),
    );
  }
}
